src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -38,28 +38,142 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
+                        token: token);
+                },
+                (actions) =>
+                {
+                    return ExecuteActionsAsync(actions, uiService.ProgressWindow, userAction, token);
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Perform the multi-package update action.
+        /// </summary>       
+        /// <remarks>This needs to be called from a background thread. It may hang on the UI thread.</remarks>
+        public async Task PerformUpdateAsync(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
+            DependencyObject windowOwner,
+            CancellationToken token)
+        {
+            await PerformActionImplAsync(
+                uiService,
+                () =>
+                {
+                    return ResolveActionsForUpdate(
+                        uiService,
+                        packagesToUpdate,
+                        token);
+                },
+                async (actions) =>
+                {
+                    foreach (var projectActions in actions.GroupBy(action => action.Item1))
+                    {
+                        await _packageManager.ExecuteNuGetProjectActionsAsync(
+                           projectActions.Key,
+                           projectActions.Select(action => action.Item2),
+                           uiService.ProgressWindow,
+                           token);
+                    }
+                },
+                windowOwner,
+                token);
+        }
+
+        /// <summary>
+        /// Calculates the list of actions needed to perform packages updates.
+        /// </summary>
+        /// <param name="uiService">ui service.</param>
+        /// <param name="packagesToUpdate">The list of packages to update.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns>The list of actions.</returns>
+        private async Task<IReadOnlyList<Tuple<NuGetProject, NuGetProjectAction>>> ResolveActionsForUpdate(
+            INuGetUI uiService,
+            List<PackageIdentity> packagesToUpdate,
+            CancellationToken token)
+        {
+            var resolvedActions = new List<Tuple<NuGetProject, NuGetProjectAction>>();
+
+            foreach (var project in uiService.Projects)
             {
-                uiService.ShowProgressDialog(windowOwner);
+                var installedPackages = await project.GetInstalledPackagesAsync(token);
+                HashSet<string> packageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                foreach (var p in installedPackages)
+                {
+                    packageIds.Add(p.PackageIdentity.Id);
+                }
+                var packagesToUpdateInProject = packagesToUpdate.Where(
+                    package => packageIds.Contains(package.Id)).ToList();
```

Comment:
We need to filter out packages from packagesToUpdate that are not installed in the current project. Otherwise, we'll incorrectly install a package that is not installed before.